server/src/main/java/org/apache/cassandra/sidecar/acl/authorization/RoleBasedAuthorizationProvider.java

```diff
@@ -0,0 +1,100 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.sidecar.acl.authorization;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.ext.auth.User;
+import io.vertx.ext.auth.authorization.Authorization;
+import io.vertx.ext.auth.authorization.AuthorizationProvider;
+import io.vertx.ext.web.handler.HttpException;
+import org.apache.cassandra.sidecar.acl.IdentityToRoleCache;
+
+import static org.apache.cassandra.sidecar.utils.AuthUtils.extractIdentities;
+
+/**
+ * Provides authorizations based on user's role. Extracts permissions user holds from Cassandra's
+ * system_auth.role_permissions table and from Sidecar's sidecar_internal.role_permissions_v1 table and sets
+ * them in user.
+ */
+public class RoleBasedAuthorizationProvider implements AuthorizationProvider
+{
+    private final IdentityToRoleCache identityToRoleCache;
+    private final RoleAuthorizationsCache roleAuthorizationsCache;
+
+    public RoleBasedAuthorizationProvider(IdentityToRoleCache identityToRoleCache,
+                                          RoleAuthorizationsCache roleAuthorizationsCache)
+    {
+        this.identityToRoleCache = identityToRoleCache;
+        this.roleAuthorizationsCache = roleAuthorizationsCache;
+    }
+
+    @Override
+    public String getId()
+    {
+        return "RoleBasedAccessControl";
+    }
+
+    @Override
+    public void getAuthorizations(User user, Handler<AsyncResult<Void>> handler)
+    {
+        getAuthorizations(user).onComplete(handler);
+    }
+
+    @Override
+    public Future<Void> getAuthorizations(User user)
+    {
+        List<String> identities = extractIdentities(user);
+
+        if (identities.isEmpty())
+        {
+            return Future.failedFuture(new HttpException(HttpResponseStatus.FORBIDDEN.code(),
+                                                         "Missing client identities"));
+        }
+
+        Set<Authorization> authorizations = new HashSet<>();
+        for (String identity : identities)
+        {
+            String role = identityToRoleCache.get(identity);
+            if (role == null)
+            {
+                continue;
+            }
+            // when entries in cache are not found, null is returned. We can not add null in user.authorizations()
+            authorizations.addAll(Optional.ofNullable(roleAuthorizationsCache.getAuthorizations(role))
+                                          .orElse(Collections.emptySet()));
```

Comment:
```suggestion
            Optional.ofNullable(roleAuthorizationsCache.getAuthorizations(role)).ifPresent(authorizations::addAll);
```
If done this way, no need to invoke authorizations.addAll with emptySet.

Reply:
thanks